keps/sig-api-machinery/2839-in-use-protection/README.md

```diff
@@ -0,0 +1,723 @@
+- Provide a generic mechanism that prevents a resource from being updated when it shouldn't be updated
+
+## Proposal
+
+A new field `Liens` to mark the object not to be deleted is introduced in object Metadata.
+Deletion requests for resources with non-empty `Liens` will be blocked by a newly introduced validation in api-server.
+
+### User Stories (Optional)
+
+#### Story 1
+
+Another controller to protect Secret from deletion watches all Secrets and their potential user objects, like `Pod`, `PersistentVolumes`, and `VolumeSnapshotContent`. Once it find the Secret is used by one of the objects, it updates the Secret's `Liens` and ask this mechanism to block deletion request for the secret.
+
+#### Story 2
+
+A user sets `Liens` to an object to be protected by this mechanism. For example, when a user knows that `Namespace` X shouldn't be erroneously deleted while `Deployment` Y exists for his/her particular use case, the user can set such a protection.
+
+<!--
+### Notes/Constraints/Caveats (Optional)
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+Most of the design ideas come from [here](https://github.com/kubernetes/kubernetes/issues/10179).
+Users or controllers can add a string to `Liens` to ask to protect a resource.
+A newly introduced validation in api-server will reject deletion requests for resources with non-empty `Liens`.
+
+`Liens` is defined as key-value pairs of string and list of struct, where key is used to identify users or controllers and value is used to describe the purpose of the use.
+Multiple users or controllers can add their `Liens` for their own purpose, and they can remove their own `Liens` when it is no longer needed.
+Deletion requests of the resource are blocked until the last `Liens` for the resource is removed
+(Just to be clear, the difference between `Liens` and `Finalizers` is that `Liens` blocks the deletion request itself, while `Finalizers` blocks the deletion to be completed).
+
+To make multiple users and controllers share the `Liens` field properly, they should include their identity to the `Liens` string and they should only delete their own `Liens` strings.
+
+A PoC implementation for in-use protection can be found, [here](https://github.com/mkimuram/kubernetes/commits/lien).
+Also, how it can be consumed by Secret protection controller can be found, [here](https://github.com/mkimuram/secret-protection/tree/lien).
+
+### API Changes
+
+`Liens` is added to `ObjectMeta`.
+
+```go
+Liens map[string][]Lien
+```
+
+Where `Lien` is defined as follows:
+
+```go
+type Lien struct {
+	LienID string
```

Comment:
I'm not conviced we need all this - can you justify it all?

Reply:
+1

Also - my comment below about growing the object indefinitely is somewhat related to it.

Reply:
yes - it seems hard to make this list bounded.  e.g. if 1000 pods depend on a secret, will we allow 1000 liens?  How about 10,000?  Those numbers are not unrealistic.

Reply:
As for the secret protection use case, all PVCs that are created from the same StorageClass may reference to the same single Secret, if the StorageClass reference to a single specific Secret. If the number of such PVCs become 1,000, it easily happens.

Reply:
Changed back to slice-of-strings Liens and restrict it to per-controller/per-user.